### PR TITLE
Add window geometry detection/Enhanced Idle, stock

### PR DESCRIPTION
**Pull request by multimokia**
# Overhauls Windowreacts framework:
- I have refactored a lot of windowreacts to be more efficient, we're no longer running needless OS checks anymore.
- I have added API funcs to get the absolute positions of the mouse cursor and MAS window, and functions to compare the two (left of MAS window, right of MAS window, etc)
- I have also built some new changes to the sprite generator, allowing us to generate sprites where Monika will follow the mouse cursor, and implemented some into our idle sprites.

# Testing:
- Verify windowreacts still work
- Verify idle looks good
- Verify no issues in sprite generation
- [ ] Linux Testing
- [ ] Windows Testing
- [ ] Mac Testing

### CODE REVIEW

File: Monika After Story/game/zz_windowutils.rpy
Comment (aldoram5) [inline Monika After Story/game/zz_windowutils.rpy:301]: This one makes notifications crash as renpy.subsitute only returns a single string and we need two to properly format this one.
Something like this should do if you need to do the substitution on the title too.
```
os.system('osascript -e \'display notification "{0}" with title "{1}"\''.format(renpy.substitute(body), renpy.substitute(title)))
```

**Review by aldoram5 — CHANGES_REQUESTED** [review]
Done with the tests on macOS Catalina, notifications are broken, check the comment I left on `zz_windowutils.rpy`.

Everything else works fine. I took a quick look at the code and looks good to me, but it's been a while since I've reviewed for MAS so take that with a grain of salt.

Extra note: Apple is changing to use a ARM based processors, the M1 being the newest and MacBooks with that may have even more issues, can't test those right now since I don't have an M1 MacBook yet. Same reason why I'm testing on Catalina instead of Big Sur.

**Review by aldoram5 — APPROVED** [review]

**Review by Booplicate — CHANGES_REQUESTED** [review]
Some optimization :mlem_boop:
My changes don't affect mac users in a different way than win/linux users, no need to test this again unless you want to, @aldoram5.

File: Monika After Story/game/updates.rpy
Comment (Booplicate) [inline Monika After Story/game/updates.rpy]: I'd delete the `rpyc`'s too.

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8864]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("1eua", weight=70),
                    MASMoniIdleExp("1eua_follow", weight=30)# 30% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.AFFECTIONATE, mas_aff.AFFECTIONATE),
                weight=94,
                tag="aff_exps"
            ),
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8869]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("1eua", weight=None),
                    MASMoniIdleExp("1eua_follow", weight=None)# 50% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.ENAMORED, mas_aff.ENAMORED),
                weight=75,
                tag="enam_exps"
            ),
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8871]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("5esu", weight=None),
                    MASMoniIdleExp("5esu_follow", weight=None)# 50% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.ENAMORED, mas_aff.ENAMORED),
                weight=11,
                tag="enam_exps"
            ),
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8876]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("1eua", weight=30),
                    MASMoniIdleExp("1eua_follow", weight=70)# 70% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.LOVE, None),
                weight=50,
                tag="love_exps"
            ),
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8878]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("5esu", weight=30),
                    MASMoniIdleExp("5esu_follow", weight=70)# 70% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.LOVE, None),
                weight=26,
                tag="love_exps"
            ),
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8882]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("5eubla", weight=30),
                    MASMoniIdleExp("5eubla_follow", weight=70)# 70% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.LOVE, None),
                weight=5,
                tag="love_exps"
            ),```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8881]:
```suggestion
```

File: Monika After Story/game/sprite-chart.rpy
Comment (Booplicate) [inline Monika After Story/game/sprite-chart.rpy:8884]:
```suggestion
            MASMoniIdleExpRngGroup(
                [
                    MASMoniIdleExp("5eubsa", weight=30),
                    MASMoniIdleExp("5eubsa_follow", weight=70)# 70% to follow
                ],
                max_uses=1,
                aff_range=(mas_aff.LOVE, None),
                weight=2,
                tag="love_exps"
            )
```

File: Monika After Story/game/updates.rpy
Comment (Booplicate) [inline Monika After Story/game/updates.rpy:381]: This `rpyc`, too.
```suggestion
        mas_utils.trydel(os.path.join(renpy.config.gamedir, "zz_windowreacts.rpy").replace('\\', '/'))
        mas_utils.trydel(os.path.join(renpy.config.gamedir, "zz_windowreacts.rpyc").replace('\\', '/'))
```

File: Monika After Story/game/script-windowreacts.rpy
Comment (Booplicate) [inline Monika After Story/game/script-windowreacts.rpy:121]: This one is missing a rule, can add it in your PR since you move wrs anyway.
```suggestion
            rules={
                "notif-group": "Window Reactions",
                "skip alert": None,
                "keep_idle_exp": None
            },
```

**Review by Booplicate — CHANGES_REQUESTED** [review]
I just remembered that this may make a good use of `MASMoniIdleExpRngGroup`. It'd a bit more cleaner and the exps would have proper weights. I left a comment with the weights I set, feel free to adjust if you wish.

**Review by Booplicate — APPROVED** [review]

**Review by Legendkiller21 — APPROVED** [review]

**Review by aldoram5 — DISMISSED** [review]

File: Monika After Story/game/zz_windowutils.rpy
Comment (Booplicate) [inline Monika After Story/game/zz_windowutils.rpy:344]:
```suggestion
        return win32gui.GetWindowRect(hwnd)
```

File: Monika After Story/game/zz_windowutils.rpy
Comment (Booplicate) [inline Monika After Story/game/zz_windowutils.rpy:315]:
```suggestion
                cur_pos = win32gui.GetCursorPos()
```

File: Monika After Story/game/zz_windowutils.rpy
Comment (Booplicate) [inline Monika After Story/game/zz_windowutils.rpy:157]: I don't like how we raise a base exception here, we can do it in a better way.
```suggestion
            if renpy.config.window_title in win32gui.GetWindowText(hwnd) and not hwnd_list:
                hwnd_list.append(hwnd)
                return True

        hwnd_list = list()
        win32gui.EnumWindows(checkMASWindow, hwnd_list)
        return hwnd_list[0] if hwnd_list else None
```

File: Monika After Story/game/zz_windowutils.rpy
Comment (Booplicate) [inline Monika After Story/game/zz_windowutils.rpy:98]: To gracefully exit from any exception.
```suggestion
            except:
```

**Review by Booplicate — CHANGES_REQUESTED** [review]
Thanks to Legend, we found some critical issues.

**Comment by multimokia** [comment]
Keeping `Don't Merge` until #6913 is merged, as update script is necessary.

**Review by aldoram5 — COMMENTED** [review]
Tested again, this time on an Apple Silicon Mac and fortunately it all works, so MAS is able to run. 
During normal gameplay it worked well under my tests. However, when I tried running `SPRITE PROG POINT TEST` game crashed and wasn't able to recover from it.
While the crash there may not affect in any way players, I rather let you guys know about it, if the test is obsolete now it may be a good time to remove it or maybe check if it can make other stuff crash in the future.
Part of the traceback:
```

I'm sorry, but an uncaught exception occurred.

While running game code:
  File "game/script-ch30.rpy", line 1474, in script call
    call call_next_event from _call_call_next_event_1
  File "game/event-handler.rpy", line 2516, in script call
    call expression event_label from _call_expression
  File "game/dev/dev_sprites.rpy", line 233, in script
    call screen mas_gen_scrollable_menu(top_level_menu, store.mas_moods.MOOD_AREA, store.mas_moods.MOOD_XALIGN, returner)
  File "renpy/common/000statements.rpy", line 466, in execute_call_screen
    args, kwargs = a.evaluate()
  File "game/dev/dev_sprites.rpy", line 233, in <module>
    call screen mas_gen_scrollable_menu(top_level_menu, store.mas_moods.MOOD_AREA, store.mas_moods.MOOD_XALIGN, returner)
AttributeError: 'StoreModule' object has no attribute 'MOOD_AREA'

-- Full Traceback ------------------------------------------------------------

Full traceback:
  File "game/script-ch30.rpy", line 1474, in script call
    call call_next_event from _call_call_next_event_1
  File "game/event-handler.rpy", line 2516, in script call
    call expression event_label from _call_expression
  File "game/dev/dev_sprites.rpy", line 233, in script
    call screen mas_gen_scrollable_menu(top_level_menu, store.mas_moods.MOOD_AREA, store.mas_moods.MOOD_XALIGN, returner)
  File "/Users/aldo/projects/renpy/old_renpy/renpy/ast.py", line 1706, in execute
    self.call("execute")
  File "/Users/aldo/projects/renpy/old_renpy/renpy/ast.py", line 1724, in call
    return renpy.statements.call(method, parsed, *args, **kwargs)
  File "/Users/aldo/projects/renpy/old_renpy/renpy/statements.py", line 145, in call
    return method(parsed, *args, **kwargs)
  File "renpy/common/000statements.rpy", line 466, in execute_call_screen
    args, kwargs = a.evaluate()
  File "/Users/aldo/projects/renpy/old_renpy/renpy/ast.py", line 184, in evaluate
    args.append(renpy.python.py_eval(v, locals=scope))
  File "/Users/aldo/projects/renpy/old_renpy/renpy/python.py", line 1749, in py_eval
    return py_eval_bytecode(code, globals, locals)
  File "/Users/aldo/projects/renpy/old_renpy/renpy/python.py", line 1743, in py_eval_bytecode
    return eval(bytecode, globals, locals)
  File "game/dev/dev_sprites.rpy", line 233, in <module>
    call screen mas_gen_scrollable_menu(top_level_menu, store.mas_moods.MOOD_AREA, store.mas_moods.MOOD_XALIGN, returner)
AttributeError: 'StoreModule' object has no attribute 'MOOD_AREA'

Darwin-20.2.0-x86_64-i386-64bit
Ren'Py 6.99.12.4.2187
Monika After Story 0.11.9

```
If this is a known issue or if it's safe to ignore it let me know and I'll approve the changes.

**Comment by ThePotatoGuy** [comment]
sprite prog point tests are important for sprite packs, so you should definitely update these, multi. I think it's just a matter of using the menu areas from styles.rpy here.

File: Monika After Story/game/event-handler.rpy
Comment (ThePotatoGuy) [inline Monika After Story/game/event-handler.rpy:2507]: can you alias `display_notif` instead of just deleting it? Especially since reacts has been out for a while now.

**Review by ThePotatoGuy — CHANGES_REQUESTED** [review]

File: Monika After Story/game/dev/dev_sprites.rpy
Comment (Booplicate) [inline Monika After Story/game/dev/dev_sprites.rpy:233]: I think this needs `mas_ui.SCROLLABLE_MENU_MEDIUM_AREA` instead.

File: Monika After Story/game/dev/dev_sprites.rpy
Comment (Booplicate) [inline Monika After Story/game/dev/dev_sprites.rpy:245]: Same here.

**Review by Booplicate — CHANGES_REQUESTED** [review]

**Review by Booplicate — APPROVED** [review]

**Review by aldoram5 — APPROVED** [review]
No crashes nor anything weird so far on my tests on macOS

**Review by Legendkiller21 — APPROVED** [review]

**Review by ThePotatoGuy — APPROVED** [review]

**Pull request merged**
